poco: 1.11.1 -> 1.12.4

tomodachi94 opened the pull request:
###### Description of changes

Changelog: https://github.com/pocoproject/poco/blob/devel/CHANGELOG

This commit also changes `src` to pull directly from GitHub.

Related: #232337

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tomodachi94 committed:
poco: 1.11.1 -> 1.12.4

Changelog: https://github.com/pocoproject/poco/blob/devel/CHANGELOG

This commit also changes `src` to pull directly from GitHub.

tomodachi94 committed:
poco: adopt package

tomodachi94 commented:
This PR fixes a security vulnerability (https://github.com/pocoproject/poco/issues/3814).

risicle commented:
@ofborg build mumble
@ofborg build pothos

Appear broken by this, nixos x86_64.

FWIW, I don't think we're affected by that security vulnerability because we replace their bundled libraries with our own (via `-DPOCO_UNBUNDLED=ON`). This is one of the main reasons we do that - we will have had those vulnerabilities fixed in our versions of those libraries ages ago.

tomodachi94 commented:
> @\\ofborg build mumble @\\ofborg build pothos
> 
> Appear broken by this, nixos x86_64.

Odd. Here's the `tail` of the build logs of `mumble` and `pathos`, it looks like the library headers aren't being added to the Store's closure.

<details><summary>Build logs</summary>
<p>

## `mumble`
```
CMake Error in src/mumble/CMakeLists.txt:
  Imported target "Poco::XML" includes non-existent path

    "/nix/store/7qrk9wmdfmm129p9y4fydr70rd59bmpk-poco-1.12.4/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in src/mumble/CMakeLists.txt:
  Imported target "Poco::XML" includes non-existent path

    "/nix/store/7qrk9wmdfmm129p9y4fydr70rd59bmpk-poco-1.12.4/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

## `pathos`

```
CMake Error in lib/CMakeLists.txt:
  Imported target "Poco::Foundation" includes non-existent path

    "/nix/store/7qrk9wmdfmm129p9y4fydr70rd59bmpk-poco-1.12.4/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.

```

</p>
</details>

<details><summary>Result of `ls /nix/store/7qrk9wmdfmm129p9y4fydr70rd59bmpk-poco-1.12.4`</summary>
<p>

```sh
$ ls /nix/store/7qrk9wmdfmm129p9y4fydr70rd59bmpk-poco-1.12.4
bin  lib
```

</p>
</details>

tomodachi94 commented:
@risicle any ideas?

tomodachi94 commented:
@infinisil @felixsinger I believe you two are the maintainers of the `mumble` package. Do either of you know how we could fix this?

(`pathos` is an orphan; let's hope the fix is similar for both 🤞)

SuperSandro2000 commented on pkgs/development/libraries/poco/default.nix:
```diff
@@ -1,38 +1,21 @@
-    })
-  ];
+  src = fetchFromGitHub {
+      owner = "pocoproject";
+      repo = "poco";
+      sha256 = "sha256-gQ97fkoTGI6yuMPjEsITfapH9FSQieR8rcrPR1nExxc=";
+      rev = "poco-${version}-release";
+    };
```
```suggestion
    owner = "pocoproject";
    repo = "poco";
    rev = "poco-${version}-release";
    hash = "sha256-gQ97fkoTGI6yuMPjEsITfapH9FSQieR8rcrPR1nExxc=";
  };
```

risicle commented:
The error is:

```
CMake Error in src/mumble/CMakeLists.txt:
  Imported target "Poco::XML" includes non-existent path

    "/nix/store/7qrk9wmdfmm129p9y4fydr70rd59bmpk-poco-1.12.4/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

The thing about multi-output derivations is they don't always work perfectly without intervention. It's not uncommon to have to do some extra fixup work to correct references.

Notice that in the new `dev` output, in `$dev/lib/cmake/Poco/PocoXMLTargets.cmake` it declares:

```
set_target_properties(Poco::XML PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "POCO_UNBUNDLED"
  INTERFACE_INCLUDE_DIRECTORIES "/nix/store/dzd9c6wwa27r1ak356g0bcvxjak26alh-poco-1.11.1-dev/include;/nix/store/dzd9c6wwa27r1ak356g0bcvxjak26alh-poco-1.11.1-dev/include"
  INTERFACE_LINK_LIBRARIES "Poco::Foundation;EXPAT::EXPAT"
)
```

whereas in the new one it declares

```
set_target_properties(Poco::XML PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "POCO_UNBUNDLED"
  INTERFACE_INCLUDE_DIRECTORIES "/nix/store/59jni2silzagbdyf9qqv1r1a97nrwbbj-poco-1.12.4-dev/include;${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "Poco::Foundation;EXPAT::EXPAT"
)
```

where `_IMPORT_PREFIX` is

```
# The installation prefix configured by this project.
set(_IMPORT_PREFIX "/nix/store/7qrk9wmdfmm129p9y4fydr70rd59bmpk-poco-1.12.4")
```

i.e. `$out`. This is probably what's making it try to look up `$out/include`. So you could figure out what changed to cause this *or* you could resort to hacking it in e.g. a `postInstall` hook.

tomodachi94 commented:
> This is probably what's making it try to look up `$out/include`. So you could figure out what changed to cause this _or_ you could resort to hacking it in e.g. a `postInstall` hook.

After further research, I suspect something similar to https://github.com/libexpat/libexpat/issues/501 is occuring in Poco; the applicable patch in Nixpkgs is this:

```nix
  # CMake files incorrectly calculate library path from dev prefix
  # https://github.com/libexpat/libexpat/issues/501
  postFixup = ''
    substituteInPlace $dev/lib/cmake/expat-${version}/expat-noconfig.cmake \
      --replace "$"'{_IMPORT_PREFIX}' $out
  '';

```

I'll test the Poco package with a similar patch.

tomodachi94 commented:
> After further research, I suspect something similar to [libexpat/libexpat#501](https://github.com/libexpat/libexpat/issues/501) is occuring in Poco; the applicable patch in Nixpkgs is this:
> 
> ```nix
>   # CMake files incorrectly calculate library path from dev prefix
>   # https://github.com/libexpat/libexpat/issues/501
>   postFixup = ''
>     substituteInPlace $dev/lib/cmake/expat-${version}/expat-noconfig.cmake \
>       --replace "$"'{_IMPORT_PREFIX}' $out
>   '';
> ```
> 
> I'll test the Poco package with a similar patch.

This wasn't working, so I turned off the `dev` output. It shouldn't affect the closure size too much.

On the plus side, `mumble` and `pothos` now build.

risicle commented:
```
  postFixup = ''
    grep -rlF INTERFACE_INCLUDE_DIRECTORIES "$dev/lib/cmake/Poco" | while read -r f; do
      substituteInPlace "$f" \
        --replace "$"'{_IMPORT_PREFIX}/include' ""
    done
  '';
```

addresses it.

`pothos` itself needs an extra patch to handle the poco 1.12 API changes:

```
    (fetchpatch {
      name = "poco-1.12-compat.patch";
      url = "https://github.com/pothosware/PothosCore/commit/092d1209b0fd0aa8a1733706c994fa95e66fd017.patch";
      hash = "sha256-bZXG8kD4+1LgDV8viZrJ/DMjg8UvW7b5keJQDXurfkA=";
    })
```

from https://github.com/pothosware/PothosCore/pull/242

tomodachi94 committed:
pothos: patch poco API calls

tomodachi94 commented:
> ```
>   postFixup = ''
>     grep -rlF INTERFACE_INCLUDE_DIRECTORIES "$dev/lib/cmake/Poco" | while read -r f; do
>       substituteInPlace "$f" \
>         --replace "$"'{_IMPORT_PREFIX}/include' ""
>     done
>   '';
> ```
> 
> addresses it.
> 
> `pothos` itself needs an extra patch to handle the poco 1.12 API changes:
> 
> ```
>     (fetchpatch {
>       name = "poco-1.12-compat.patch";
>       url = "https://github.com/pothosware/PothosCore/commit/092d1209b0fd0aa8a1733706c994fa95e66fd017.patch";
>       hash = "sha256-bZXG8kD4+1LgDV8viZrJ/DMjg8UvW7b5keJQDXurfkA=";
>     })
> ```
> 
> from [pothosware/PothosCore#242](https://github.com/pothosware/PothosCore/pull/242)

Ohmygod thank you, I've been poring over this all week. I've added both of your patches into this PR. Again, thank you so much!

risicle reviewed (approved):
`nixpkgs-review` happy, macos 10.15 & nixos x86_64
